Improve coming soon store page matching

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/53129

This PR is to improve store page detection logic. It's challenging to use url matching to detect store pages. I think it's better to use native WordPress methods (e.g. `is_page()`, `is_post_type_archive()`, `is_singular()`, etc.) to detect store pages, which are more reliable and easier to maintain, so I introduced a new method `is_current_page_store_page()` to determine if the current page is a store page.

The only downside is that we can only use this method when WP has set up the query, typically during or after the `parse_query` or `template_redirect` action hooks. This causes an issue when we need to show the coming soon page in non-FSE themes. In [this PR](https://github.com/woocommerce/woocommerce/pull/54624), I fixed the "Coming Soon Social Links" template part for classic themes by moving the initialization of the Block templates to the `after_setup_theme` hook. The initialization is only needed when we need to show the coming soon page in non-FSE themes, but since we can't use the `is_current_page_store_page()` method in `after_setup_theme`, I changed it to initialize the class when coming soon mode is active, regardless of the page. I think this's a good compromise since it [doesn't cause any impact on functionality when the block template](https://github.com/woocommerce/woocommerce/issues/46461) is initialized for non-coming soon pages.

Changes:

- Add `is_current_page_store_page()` method, which use WP native methods to detect store pages.
- Deprecate the `is_store_page()` method with a notice
- Add test cases for the new method

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


**1. Basic Store Page Detection**
1. Create a new JN site with this branch.
2. Install and activate `WooCommerce Smooth Generator` plugin.
3. Generate some products using `WooCommerce Smooth Generator` plugin.
4. Navigate to `WooCommerce -> Settings -> Site Visibility` and choose `Coming soon with Restrict to store pages only`.
5. Open the site in incognito mode and navigate to any store page (Shop, Cart, Checkout, Product, Category, etc)
6. Confirm that the coming soon page is shown.

**2. URL Pattern Collision Prevention**
1. Create a new page from `Pages -> Add a new page`
2. Set the page title to `Shopping URL` and confirm the permalink is set to `shopping-url`.
3. Open the page in an incognito session.
4. Confirm that the coming soon page is NOT shown.
5. Set the link to `products-demo` and save the page
6. Open the page in an incognito session.
7. Confirm that the coming soon page is NOT shown.

**3. Plain Permalink Structure with Query Parameters**
1. Go to `Settings > Permalinks`
2. Set permalink settings to `Plain`
3. In an incognito browser, go to Cart page
4. Confirm that the coming soon page is shown.
5. Add `&tag=test` to the URL and verify coming soon page still shows

**4. Shop Page Detection with Plain Permalinks**
1. Open an incognito session and navigate to the frontend.
2. Click `Shop` from the menu.
3. Confirm the coming soon page is rendered.

**5. Unassigned Shop Page Handling**
1. Go to /wp-admin/admin.php?page=wc-settings&tab=products
2. Unassign the shop page and save changes
3. Open a incognito window and go to the /shop page
4. Verify that the coming soon page is displayed

***

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Refactor WCAdminHelper to improve store page detection

- Replace `is_store_page()` with `is_current_page_store_page()` method
- Simplify store page detection logic using WordPress conditional tags
- Add support for product pages, categories, and tags
- Deprecate the old method with a notice
- Improve test coverage for store page detection

chihsuan committed:
Add changelog

chihsuan committed:
Fix lint

chihsuan committed:
Update order

chihsuan committed:
Improve Coming Soon mode and store page detection

- Refactor ComingSoonRequestHandler to skip initialization when site is live
- Update ComingSoonHelper method description for clarity
- Simplify initialization logic in ComingSoonRequestHandler
- Update WCAdminHelper method documentation for is_current_page_store_page()

chihsuan committed:
Update tests

ilyasfoo commented on plugins/woocommerce/src/Admin/WCAdminHelper.php:
```diff
@@ -181,116 +148,35 @@ public static function is_store_page( $url = '' ) {
+			is_tax( 'product_tag' ) ||
+			is_tax( 'product_cat' ) ||
+			is_tax( 'product_brand' ) ||
+			is_tax( 'product_variation' )
```
Could we use [`is_product_taxonomy`](https://github.com/woocommerce/woocommerce/blob/5d7d14cd7ab7c37a0834d7f1b4192bf2c8000b97/plugins/woocommerce/includes/wc-conditional-functions.php#L47) here instead? I think we can leverage this as it should also capture custom product taxonomies by extensions. It gives the following taxonomies in a fresh WC install:

```
wp> var_dump(get_object_taxonomies( 'product' ));
array(6) {
  [0]=>
  string(13) "product_brand"
  [1]=>
  string(12) "product_type"
  [2]=>
  string(18) "product_visibility"
  [3]=>
  string(11) "product_cat"
  [4]=>
  string(11) "product_tag"
  [5]=>
  string(22) "product_shipping_class"
}
```

However, I'm unsure why `product_variation` isnt listed. As far as I can tell, there's no product page that uses `product_variation` taxonomy. Let me know what you think!

chihsuan replied:
> Could we use [is_product_taxonomy](https://github.com/woocommerce/woocommerce/blob/5d7d14cd7ab7c37a0834d7f1b4192bf2c8000b97/plugins/woocommerce/includes/wc-conditional-functions.php#L47) here instead? I think we can leverage this as it should also capture custom product taxonomies by extensions. It gives the following taxonomies in a fresh WC install:

Yeah, that's a great suggestion! I've updated the PR to use is_product_taxonomy. a04b4eda6baeed7da572d87eca52d308ac8391be

> However, I'm unsure why product_variation isnt listed. As far as I can tell, there's no product page that uses product_variation taxonomy. Let me know what you think!

Looked into it, and product_variation is indeed not a taxonomy. I've removed it from the list.

ilyasfoo commented on plugins/woocommerce/src/Admin/WCAdminHelper.php:
```diff
@@ -322,4 +220,21 @@ private static function get_normalized_url_path( $url ) {
+		// Initialize query vars if they haven't been set.
+		if ( empty( $wp->query_vars ) || empty( $wp->request ) ) {
+			$wp->parse_request();
+
```
Nit
```suggestion
```

ilyasfoo commented on plugins/woocommerce/src/Internal/ComingSoon/ComingSoonHelper.php:
```diff
@@ -30,11 +30,9 @@ public function is_store_coming_soon(): bool {
-	 * @param string $url The URL to check.
+	 * Return true if the current page should be shown in coming soon mode.
 	 */
-	public function is_url_coming_soon( string $url ): bool {
```
`is_url_coming_soon` was previously a public function, do we need to deprecate it? I highly doubt it was used by 3pd though, no results from "all-plugins" repository

chihsuan replied:
Since it's under Internal namespace, I think we don't need to deprecate it. Thanks for checking!

ilyasfoo commented on plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php:
```diff
@@ -36,28 +36,30 @@ class ComingSoonRequestHandler {
 	final public function init( ComingSoonHelper $coming_soon_helper ) {
 		$this->coming_soon_helper = $coming_soon_helper;
+		// Skip if the site is live.
+		if ( $this->coming_soon_helper->is_site_live() ) {
```
Nice optimization!

ilyasfoo reviewed (commented):
Amazing work here, @chihsuan, thank you!

moon0326 reviewed (approved):
Thank you so much for this PR! Looks much cleaner and it's easier to understand.

Tested various scenarios including the previous bugs with shop page detection. Everything worked as expected 👍 🚀

chihsuan committed:
Update plugins/woocommerce/src/Admin/WCAdminHelper.php

Co-authored-by: Ilyas Foo <foo.ilyas@gmail.com>

chihsuan committed:
Simplify WCAdminHelper product taxonomy page detection

- Replace multiple `is_tax()` checks with WordPress's `is_product_taxonomy()` method
- Improve code readability and maintainability

ilyasfoo reviewed (approved):
Nice, thanks! LGTM 🚢
